File: AOC_2023/Day 14/14.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func initialize(filename string) ([][] rune, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	var grid [][] rune
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, []rune(scanner.Text()))
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return grid, nil
}

func simulate(direction rune, grid [][]rune) {
    n, m, k := len(grid), len(grid[0]), 0
	switch direction {
	case 'N':
		for i := 1; i < n; i++ {
			for j := 0; j < m; j++ {
				if grid[i][j] == 'O'{
					for k = i; k > 0 && grid[k-1][j] == '.'; k-- {}
					grid[i][j], grid[k][j] = '.', 'O'
				}
			}
		}
	case 'W':
		for i := 0; i < n; i++ {
			for j := 1; j < m; j++ {
				if grid[i][j] == 'O'{
					for k = j; k > 0 && grid[i][k-1] == '.'; k-- {}
					grid[i][j], grid[i][k] = '.', 'O'
				}
			}
		}
	case 'S':
		for i := n-2; i >= 0; i-- {
			for j := 0; j < m; j++ {
				if grid[i][j] == 'O'{
					for k = i; k < n-1 && grid[k+1][j] == '.'; k++ {}
					grid[i][j], grid[k][j] = '.', 'O'
				}
			}
		}
	case 'E':
		for i := 0; i < n; i++ {
			for j := m-2; j >= 0; j-- {
				if grid[i][j] == 'O'{
					for k = j; k < m-1 && grid[i][k+1] == '.'; k++ {}
					grid[i][j], grid[i][k] = '.', 'O'
				}
			}
		}

	}

}

func getLoad(grid [][]rune) int {
	result := 0
	n, m := len(grid), len(grid[0])
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == 'O' {
				result += n-i
			}
		}
	}
	return result
}

func gridToString(grid [][]rune) string {
	result := ""
	for _, row := range grid {
		result += string(row)
	}
	return result
}

func part1(grid [][] rune) {
	simulate('N', grid)
	fmt.Println("Part 1:", getLoad(grid))
}

func part2(grid [][] rune) {
	cycles,sim := 1000000000, 0
	directions := []rune{'N', 'W', 'S', 'E'}
	seen := make(map[string]int)
	for i := 1; i <= cycles; i++ {
		for _, direction := range directions {
			simulate(direction, grid)
		}
		gridString := gridToString(grid)
		if _, ok := seen[gridString]; ok {
			sim = i + (((cycles - i)/(i - seen[gridString])) * (i - seen[gridString]))
			break
		}
		seen[gridString] = i
	}
	for i := sim+1; i <= cycles; i++ {
		for _, direction := range directions {
			simulate(direction, grid)
		}
	}
	fmt.Println("Part 2:", getLoad(grid))
}

func main(){
	filename := "input.txt"
	grid, err := initialize(filename)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	copyGrid := make([][]rune, len(grid))
	for i := range grid {
		copyGrid[i] = make([]rune, len(grid[i]))
		copy(copyGrid[i], grid[i])
	}
	part1(grid)
	grid = copyGrid
	part2(grid)
}

File: AOC_2023/Day 14/14.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>

using namespace std;

class Solution
{
    public:
    int n, m;
    vector<string> grid, copyGrid;
    void init(ifstream& input){
        string line;
        while (getline(input, line))
        {
            grid.push_back(line);
        }
        copyGrid = grid;
        n = grid.size(), m = grid[0].size();
    }

    void simulate(char direction){
        if(direction == 'N'){
            for (int i = 1; i < n; i++)
            {
                for(int j = 0; j < grid[i].size(); j++){
                    if(grid[i][j] == 'O'){
                        int k = i;
                        while(k > 0 && grid[k-1][j] == '.'){
                            k--;
                        }
                        grid[i][j] = '.';
                        grid[k][j] = 'O';
                    }
                }
            }
        }
        else if(direction == 'W'){
            for(int j = 1; j < m; j++){
                for (int i = 0; i < n; i++)
                {
                    if(grid[i][j] == 'O'){
                        int k = j;
                        while(k > 0 && grid[i][k-1] == '.'){
                            k--;
                        }
                        grid[i][j] = '.';
                        grid[i][k] = 'O';
                    }
                }
            }
        }
        else if(direction == 'S'){
            for (int i = n - 2; i >= 0; i--)
            {
                for(int j = 0; j < grid[i].size(); j++){
                    if(grid[i][j] == 'O'){
                        int k = i;
                        while(k < n - 1 && grid[k+1][j] == '.'){
                            k++;
                        }
                        grid[i][j] = '.';
                        grid[k][j] = 'O';
                    }
                }
            }
        }
        else if(direction == 'E'){
            for(int j = m - 2; j >= 0; j--){
                for (int i = 0; i < n; i++)
                {
                    if(grid[i][j] == 'O'){
                        int k = j;
                        while(k < m - 1 && grid[i][k+1] == '.'){
                            k++;
                        }
                        grid[i][j] = '.';
                        grid[i][k] = 'O';
                    }
                }
            }
        }
    }

    string gridToString(){
        string result = "";
        for(string s : grid)
            result += s;
        return result;
    }

    int totalLoad(){
        int result = 0;
        for (int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == 'O'){
                    result += (n - i);
                }
            }
        }
        return result;
    }

    void part1(){
        simulate('N');
        cout<<"Part 1: "<<totalLoad()<<"\n";
    }

    void part2(){
        int cycles = 1000000000,sim;
        grid = copyGrid;
        char directions[4] = {'N', 'W', 'S', 'E'};
        unordered_map<string, int> seen;
        for(int i = 1; i <= cycles;i++){
            for(char direction : directions)
                simulate(direction);
            string gridString = gridToString();
            if(seen.count(gridString)){
                sim = i + (((cycles - i)/(i - seen[gridString])) * (i - seen[gridString]));
                break;
            }
            seen[gridString] = i;
        }
        for(int i = sim+1; i <= cycles;i++){
            for(char direction : directions)
                simulate(direction);
        }
        cout<<"Part 2: "<<totalLoad()<<"\n";
    }

};

int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}